File: apps/windows/AiUsage.Windows/Services/ClaudeSignIn.cs
using System.Diagnostics;

namespace AiUsage.Windows.Services;

internal static class ClaudeSignIn
{
    private static Process? activeProcess;

    public static void CancelActiveProcess()
    {
        if (activeProcess is { HasExited: false } process)
        {
            process.Kill(entireProcessTree: true);
        }
    }

    public static async Task<int> RenewSessionAsync(CancellationToken cancellationToken)
    {
        // Credential verification decides whether startup renewed the session
        return await ExecuteAsync(
            ["--print", "--tools=", "--no-session-persistence", "--setting-sources=",
             "--settings", "{\"disableAllHooks\":true}", "--strict-mcp-config",
             "--mcp-config", "{\"mcpServers\":{}}", "--disable-slash-commands"],
            closeInput: true,
            timeoutDuration: TimeSpan.FromSeconds(30),
            cancellationToken);
    }

    public static async Task SignInAsync(CancellationToken cancellationToken)
    {
        var status = await ExecuteAsync(["auth", "login"], closeInput: false, timeoutDuration: TimeSpan.FromMinutes(10), cancellationToken);
        if (status != 0)
        {
            throw new InvalidOperationException("Claude sign-in did not complete.");
        }
    }

    private static async Task<int> ExecuteAsync(string[] arguments, bool closeInput, TimeSpan timeoutDuration, CancellationToken cancellationToken)
    {
        var executable = FindExecutable() ?? throw new FileNotFoundException("Claude Code was not found.");
        var temporaryDirectory = closeInput ? Directory.CreateTempSubdirectory("ai-usage-claude-") : null;
        try
        {
            return await RunProcessAsync(executable, arguments, temporaryDirectory?.FullName, closeInput, timeoutDuration, cancellationToken);
        }
        finally
        {
            temporaryDirectory?.Delete(recursive: true);
        }
    }

    private static async Task<int> RunProcessAsync(string executable, string[] arguments, string? workingDirectory, bool closeInput, TimeSpan timeoutDuration, CancellationToken cancellationToken)
    {
        var startInfo = new ProcessStartInfo(executable)
        {
            UseShellExecute = false,
            CreateNoWindow = true,
            RedirectStandardInput = true,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            WorkingDirectory = workingDirectory ?? Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
        };
        if (Path.GetExtension(executable).Equals(".cmd", StringComparison.OrdinalIgnoreCase))
        {
            startInfo.FileName = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.System), "WindowsPowerShell", "v1.0", "powershell.exe");
            startInfo.Environment["AI_USAGE_CLAUDE_EXECUTABLE"] = executable;
            startInfo.ArgumentList.Add("-NoProfile");
            startInfo.ArgumentList.Add("-Command");
            var quotedArguments = arguments.Select(argument => "'" + argument.Replace("\"", "\\\"").Replace("'", "''") + "'");
            startInfo.ArgumentList.Add("& $env:AI_USAGE_CLAUDE_EXECUTABLE " + string.Join(" ", quotedArguments) + "; exit $LASTEXITCODE");
        }
        else
        {
            foreach (var argument in arguments)
            {
                startInfo.ArgumentList.Add(argument);
            }
        }

        using var process = new Process { StartInfo = startInfo };
        cancellationToken.ThrowIfCancellationRequested();
        process.Start();
        activeProcess = process;
        if (closeInput)
        {
            process.StandardInput.Close();
        }

        using var timeout = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        timeout.CancelAfter(timeoutDuration);
        var output = process.StandardOutput.BaseStream.CopyToAsync(Stream.Null, timeout.Token);
        var error = process.StandardError.BaseStream.CopyToAsync(Stream.Null, timeout.Token);
        try
        {
            await Task.WhenAll(process.WaitForExitAsync(timeout.Token), output, error);
            cancellationToken.ThrowIfCancellationRequested();

            return process.ExitCode;
        }
        catch (OperationCanceledException) when (!cancellationToken.IsCancellationRequested)
        {
            throw new TimeoutException("Claude sign-in timed out.");
        }
        finally
        {
            activeProcess = null;
            if (!process.HasExited)
            {
                process.Kill(entireProcessTree: true);
                await process.WaitForExitAsync();
            }
        }
    }

    private static string? FindExecutable()
    {
        var home = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        var paths = new[] { Path.Combine(home, ".local", "bin"), Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "npm") }
            .Concat((Environment.GetEnvironmentVariable("PATH") ?? "").Split(Path.PathSeparator, StringSplitOptions.RemoveEmptyEntries));

        return paths.SelectMany(path => new[] { Path.Combine(path, "claude.exe"), Path.Combine(path, "claude.cmd") })
            .FirstOrDefault(File.Exists);
    }
}

File: Sources/AiUsageApp/Providers/ClaudeSignIn.swift
import Darwin
import Foundation

enum ClaudeSignInError: Error, Equatable {
    case notInstalled
    case failed
    case timedOut
}

@MainActor
enum ClaudeSignIn {
    private static var activeProcess: Process?

    static func cancelActiveProcess() {
        if let process = activeProcess, process.isRunning {
            process.terminate()
            if process.isRunning {
                kill(process.processIdentifier, SIGKILL)
            }
        }
    }

    @discardableResult
    static func renewSession(executableURL: URL? = nil, timeout: Duration = .seconds(30)) async throws -> Int32 {
        // Credential verification decides whether startup renewed the session
        try await execute(
            executableURL: executableURL,
            arguments: [
                "--print", "--tools", "", "--no-session-persistence",
                "--setting-sources", "", "--settings", "{\"disableAllHooks\":true}",
                "--strict-mcp-config", "--mcp-config", "{\"mcpServers\":{}}",
                "--disable-slash-commands",
            ],
            closeInput: true,
            timeout: timeout
        )
    }

    static func signIn(executableURL: URL? = nil, timeout: Duration = .seconds(600)) async throws {
        let status = try await execute(executableURL: executableURL, arguments: ["auth", "login"], closeInput: false, timeout: timeout)
        guard status == 0 else {
            throw ClaudeSignInError.failed
        }
    }

    private static func execute(executableURL: URL?, arguments: [String], closeInput: Bool, timeout: Duration) async throws -> Int32 {
        guard let executableURL = executableURL ?? findExecutable() else {
            throw ClaudeSignInError.notInstalled
        }

        let process = Process()
        process.executableURL = executableURL
        process.arguments = arguments
        var environment = ProcessInfo.processInfo.environment
        environment["PATH"] = executableURL.deletingLastPathComponent().path + ":/opt/homebrew/bin:/usr/local/bin:" + (environment["PATH"] ?? "/usr/bin:/bin")
        process.environment = environment
        let temporaryDirectory = closeInput
            ? FileManager.default.temporaryDirectory.appendingPathComponent("ai-usage-claude-" + UUID().uuidString)
            : nil
        if let temporaryDirectory {
            try FileManager.default.createDirectory(at: temporaryDirectory, withIntermediateDirectories: true)
        }
        defer {
            if let temporaryDirectory {
                try? FileManager.default.removeItem(at: temporaryDirectory)
            }
        }
        process.currentDirectoryURL = temporaryDirectory ?? FileManager.default.homeDirectoryForCurrentUser
        let input = Pipe()
        process.standardInput = input
        process.standardOutput = FileHandle.nullDevice
        process.standardError = FileHandle.nullDevice

        if closeInput {
            try input.fileHandleForWriting.close()
        }
        try Task.checkCancellation()
        try process.run()
        activeProcess = process
        defer {
            activeProcess = nil
            try? input.fileHandleForWriting.close()
            if process.isRunning {
                process.terminate()
                if process.isRunning {
                    kill(process.processIdentifier, SIGKILL)
                }
            }
        }

        let deadline = ContinuousClock.now + timeout
        while process.isRunning {
            try Task.checkCancellation()
            guard ContinuousClock.now < deadline else {
                throw ClaudeSignInError.timedOut
            }

            try await Task.sleep(for: .milliseconds(200))
        }

        try Task.checkCancellation()
        return process.terminationStatus
    }

    private static func findExecutable() -> URL? {
        let home = FileManager.default.homeDirectoryForCurrentUser.path
        let paths = ["\(home)/.local/bin", "/opt/homebrew/bin", "/usr/local/bin"]
            + (ProcessInfo.processInfo.environment["PATH"] ?? "").split(separator: ":").map(String.init)

        return paths.map { URL(fileURLWithPath: $0).appendingPathComponent("claude") }
            .first { FileManager.default.isExecutableFile(atPath: $0.path) }
    }
}

@MainActor
struct ClaudeRecoveryClient {
    var renewSession: () async throws -> Int32
    var signIn: () async throws -> Void
    var refreshUsage: (String?, Date) async throws -> ProviderSnapshot

    static let live = ClaudeRecoveryClient(
        renewSession: { try await ClaudeSignIn.renewSession() },
        signIn: { try await ClaudeSignIn.signIn() },
        refreshUsage: { credentials, now in
            try await SharedCoreClient().refreshClaude(claudeCredentialsJSON: credentials, now: now)
        }
    )
}
